tools/fork-network/src/cli.rs

```diff
@@ -1,191 +1,723 @@
+            Ok(true)
+        }
+    }
 
-        let storage = store_opener.open_in_mode(Mode::ReadWrite).unwrap();
+    fn init(&self, near_config: &mut NearConfig, home_dir: &Path) -> anyhow::Result<()> {
+        // Open storage with migration
+        let storage = open_storage(&home_dir, near_config).unwrap();
         let store = storage.get_hot_store();
+        assert!(self.snapshot_db(store.clone(), near_config, home_dir)?);
 
         let epoch_manager =
             EpochManager::new_arc_handle(store.clone(), &near_config.genesis.config);
-        let runtime = NightshadeRuntime::from_config(
-            home_dir,
-            store.clone(),
-            &near_config,
-            epoch_manager.clone(),
-        );
-        let head = store.get_ser::<Tip>(DBCol::BlockMisc, HEAD_KEY)?.unwrap();
+        let head = store.get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?.unwrap();
         let shard_layout = epoch_manager.get_shard_layout(&head.epoch_id)?;
         let all_shard_uids = shard_layout.get_shard_uids();
+        let fork_head = get_fork_head(&all_shard_uids, store.clone())?;
+        tracing::info!(?fork_head);
 
-        let fork_head = {
-            let mut flat_head: Option<CryptoHash> = None;
-            for shard_uid in all_shard_uids {
-                let flat_storage_status = store
-                    .get_ser::<FlatStorageStatus>(DBCol::FlatStorageStatus, &shard_uid.to_bytes())?
-                    .unwrap();
-                if let FlatStorageStatus::Ready(ready) = &flat_storage_status {
-                    let flat_head_for_this_shard = ready.flat_head.hash;
-                    if let Some(hash) = flat_head {
-                        if hash != flat_head_for_this_shard {
-                            panic!("Not all shards have the same flat head. Please reset the fork, and run the node for a little longer");
-                        }
-                    }
-                    flat_head = Some(flat_head_for_this_shard);
-                } else {
-                    panic!(
-                        "Flat storage is not ready for shard {}: {:?}. Please reset the fork, and run the node for longer",
-                        shard_uid, flat_storage_status
-                    );
-                }
-            }
-            println!("Forking from the flat storage final head: {}", flat_head.unwrap());
-            flat_head.unwrap()
-        };
+        let chain =
+            ChainStore::new(store.clone(), near_config.genesis.config.genesis_height, false);
 
-        let fork_head_block = store
-            .get_ser::<near_primitives::block::Block>(
-                DBCol::Block,
-                &borsh::to_vec(&fork_head).unwrap(),
-            )?
-            .unwrap();
+        let map = chain.get_all_block_hashes_by_height(fork_head.height)?;
+        let (_, hashes) = (*map).iter().next().unwrap();
+        let block_hash = *hashes.iter().next().unwrap();
```

Comment:
these two `.iter().next()` will just result in some arbitrary block of the same height as the fork head block, right? If there are multiple blocks at that height, then the one picked is not well defined I think. Why can we not just use `fork_head.hash` where `block_hash` is used below? Especially since that is what is used to get state roots with the `get_chunk_extra()` call, while the `shard_uid` argument is built from this arbitrary block instead of the actual `fork_head`

Reply:
You are right, I got confused. The interface for getting a trie from `ShardTries` requests block hash that corresponds to post-state, i.e. `fork_head.hash` is right. However, this hash value is ignored, because `use_flat_storage` is `false`.

`block_height` I'd like to set to `fork_head.height + 1`, but it's only usage is the `height` field in the new genesis.